File: backend/src/api/v1/archive/production_groups.py
from typing import List

from fastapi import APIRouter, Depends, Query, Request, status
from sqlalchemy.orm import Session

from src.api.dependencies import RequirePermissions
from src.database import get_db
from src.models.user import User
from src.schemas.production_group import (
    ProductionGroupCreate,
    ProductionGroupResponse,
    ProductionGroupUpdate,
)
from src.services.archive import get_base_url
from src.services.auth.permissions import Permissions
from src.services.production_group import (
    create_production_group,
    delete_production_group_by_id,
    get_production_group_by_id,
    get_production_groups_list,
    update_production_group,
)

router = APIRouter()


@router.get(
    "/",
    response_model=List[ProductionGroupResponse],
    summary="Get production groups",
    description="Returns production groups, optionally including hidden groups.",
)
async def get_production_groups(
    request: Request,
    db: Session = Depends(get_db),
    public_only: bool = Query(True),
):
    base_url = get_base_url(str(request.url))
    return get_production_groups_list(db, base_url, public_only=public_only)


@router.get(
    "/{production_group_id}",
    response_model=ProductionGroupResponse,
    summary="Get a production group",
)
async def get_production_group(
    production_group_id: int,
    request: Request,
    db: Session = Depends(get_db),
):
    base_url = get_base_url(str(request.url), remove_last_segments=2)
    return get_production_group_by_id(db, production_group_id, base_url)


@router.post(
    "/",
    response_model=ProductionGroupResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create a production group",
)
async def post_production_group(
    production_group_in: ProductionGroupCreate,
    request: Request,
    db: Session = Depends(get_db),
    _: User = Depends(RequirePermissions([Permissions.ARCHIVE_CREATE])),
):
    base_url = get_base_url(str(request.url))
    return create_production_group(db, production_group_in, base_url)


@router.patch(
    "/{production_group_id}",
    response_model=ProductionGroupResponse,
    summary="Update a production group",
)
async def patch_production_group(
    production_group_id: int,
    production_group_in: ProductionGroupUpdate,
    request: Request,
    db: Session = Depends(get_db),
    _: User = Depends(RequirePermissions([Permissions.ARCHIVE_UPDATE])),
):
    base_url = get_base_url(str(request.url), remove_last_segments=2)
    return update_production_group(
        db, production_group_id, production_group_in, base_url
    )


@router.delete(
    "/{production_group_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete a production group",
)
async def delete_production_group(
    production_group_id: int,
    db: Session = Depends(get_db),
    _: User = Depends(RequirePermissions([Permissions.ARCHIVE_DELETE])),
):
    delete_production_group_by_id(db, production_group_id)

File: backend/src/api/v1/archive/production_groups.rs
use axum::{
    extract::{Query, State},
    http::StatusCode,
    Json,
};

use crate::api::dependencies::auth::CurrentUser;
use crate::errors::AppError;
use crate::schemas::production_group::{
    ProductionGroupCreate, ProductionGroupListQuery, ProductionGroupResponse,
};
use crate::services::auth::permissions::{
    check_permission, PERMISSION_ARCHIVE_READ, PERMISSION_ARCHIVE_WRITE,
};
use crate::services::auth::user::get_user_permission_names;
use crate::services::production_group::{create_production_group, list_production_groups};
use crate::AppState;

pub async fn list_production_groups_handler(
    State(state): State<AppState>,
    CurrentUser(user): CurrentUser,
    Query(query): Query<ProductionGroupListQuery>,
) -> Result<Json<Vec<ProductionGroupResponse>>, AppError> {
    let perms = get_user_permission_names(&state.db, user.id).await?;
    if !user.super_user {
        check_permission(&perms, PERMISSION_ARCHIVE_READ)?;
    }
    let result = list_production_groups(&state.db, query).await?;
    Ok(Json(result))
}

pub async fn create_production_group_handler(
    State(state): State<AppState>,
    CurrentUser(user): CurrentUser,
    Json(data): Json<ProductionGroupCreate>,
) -> Result<(StatusCode, Json<ProductionGroupResponse>), AppError> {
    let perms = get_user_permission_names(&state.db, user.id).await?;
    if !user.super_user {
        check_permission(&perms, PERMISSION_ARCHIVE_WRITE)?;
    }
    let result = create_production_group(&state.db, data).await?;
    Ok((StatusCode::CREATED, Json(result)))
}
